LuaDoc. Fixed the start position of the comment first symbol in docs

TIMONz1535 opened the pull request:
~~The positions were calculated slightly incorrectly there, because it got confused with the lengths.~~

```
---@param a string?Comment
local function new(a) end

---@param a string[]Comment
local function new(a) end
```
now its `Comment` instead of `omment`.

![изображение](https://github.com/user-attachments/assets/2fc4361d-3475-44f2-ab24-9f4375c33f67)

~~But I came across the fact that the comment texts `comment.text` do not contain `--` at the beginning, because this token is uuuh... being skipped?~~

~~That is, we were working with a string of a different length and start pos (-2). I made a variable `local textOffset = 2` so that it could be easily fixed in the future.~~

~~p.s. Maybe you should figure out why this is happening.~~

## upd. Please see the latest message for actual changes.

tomlau10 commented on script/parser/luadoc.lua:
```diff
@@ -1743,11 +1743,15 @@ local function buildLuaDoc(comment)
+        local cpos = finish - comment.start + 1
+        -- actual text start is `comment.start + 2` because of `--` for some reason
+        local textOffset = 2
+        local cstart = text:find('%S', cpos - textOffset)
```
IIUC, `comment.start|finish` are **0-based** offset relative to **the original line of text**.
For example:
```lua
---@param a string
--> start = 0, finish = 18 (i.e. the length of the original line)

  ---@param a string
--> start = 2, finish = 20 (because there are 2 spaces before the luadoc comment)

  --[[@param a string]] -- other
--> start = 2, finish = 23 (a block comment can end earlier than the whole line)
```
- therefore the original logic already has an **off-by-1** error when doing `text:find()`
because in lua string offset is **1-based**
- now you also found out that `text` (which is a local var to `comment.text`) is with the `--` prefix stripped
so there is **another** offset error when the original logic just doing `text:find('%S', finish - comment.start)`
  - `finish - comment.start` this is a **0-based** offset relative to **original line of text** (before `--` being stripped)
  - certainly it should not be directly used as offset when doing `:find()` 👍 

---

## Better readability?

With all these knowledge, I would suggest using a pair of conversion function like:
```lua
local function linePosToCommentPos(linePos)
    -- line pos is 0-based relative to original line of text before `--` prefix is removed
    -- while comment pos is 1-based and with `--` prefix removed
    return linePos + 1 - 2
end
local function commentPosToLinePos(commentPos)
    return commentPos + 2 - 1
end
-- maybe with better function names if you have other ideas
```
Then we may have the following code for better readability 🤔 
```lua
        local cstart = text:find('%S', linePosToCommentPos(finish - comment.start))
        if cstart and commentPosToLinePos(cstart) < comment.finish then
            result.comment = {
                type   = 'doc.tailcomment',
                start  = comment.start + commentPosToLinePos(cstart),
```
This also maintains almost the same logic flow as the original, only that a few extra conversion function is being called.

TIMONz1535 replied:
Thanks for the review, you're amazing. I didn't notice a mistake in the condition 🫠

However, I want to say that I mentioned the **length** for a reason. Subtracting the positions gives a **length** that is also equal to the **relative 0-based position** - but it doesn't depend on what the positions were based on 0-b or 1-b.

![изображение](https://github.com/user-attachments/assets/1f976c12-2337-48c6-8102-430bf3fb3736)

Anyway your functions are universal.

tomlau10 commented:
> But I came across the fact that the comment texts `comment.text` do not contain `--` at the beginning, because this token is uuuh... being skipped?

Regarding this question, I think it makes sense because it stores the **body** of a comment 🤔 
```lua
--foo bar
```
- the comment body is only `foo bar`, not `--foo bar`
- the prefix `--` is not part of the comment body

tomlau10 commented:
> Also the expression `text:find('%S'` seems pointless, because we're already doing `util.trim (trimTailComment)`. Maybe it should have eliminated the empty spaces comments `---@param a string? `.

After some thought, this is **not** pointless 🤔 
- `%S` means non space characters
- by using `text:find('%S', <comment start offset>)`, we are **filtering out empty comments**
  - `cstart = nil` if no comment string are found
  - then the logic will not execute the if case below
  - => therefore we can avoid generating empty `result.comment` objects
- otherwise for the case `---@param a string? `, even though `util.trim` can trim all the spaces, an useless `result.comment` will still be generated

> but we can never have the end char pos of a string `>= comment.finish`

If we can be certain that the **length** of `comment.text` is always `<= comment.finish - comment.start`,
then I agree that the `convertCommentPosToLinePos(comment, cstart) < comment.finish` check is pointless. 👍 
As you said, in this case we can never have the **converted trailing comment start pos** `>= finish`

TIMONz1535 commented:
I have redone this.

### Fix calculating the comment over `doc` not `text` in **buildLuaDoc**.

Actually we call `parseTokens(doc, ...)` and `convertTokens(doc)` so our `result` tokens is from `doc`, not `comment.text`. That's why we don't use `text` anymore.

https://github.com/LuaLS/lua-language-server/blob/9ba27a6edd0ec0c73c3d59b3b1b028a1a46138c6/script/parser/luadoc.lua#L1716-L1736

The `doc` is the string **after @ symbol**, so I made a variable `startOffset` that means an **absolute position of `@` symbol**. The `startOffset` for the **comment.short** should be `comment.start + headPos + 2 - 2` but I simplified it. Here `+2` is `--` prefix offset and `-2` is back from "1-based after `@` pos" to "0-based of `@` pos". In `cstart` calculations I replaced the `comment.start` with `startOffset`, nothing else is needed.

Later, I discovered exactly the same logic in the `semantic-tokens.lua`, and decided to synchronize them.

https://github.com/LuaLS/lua-language-server/blob/9ba27a6edd0ec0c73c3d59b3b1b028a1a46138c6/script/core/semantic-tokens.lua#L906-L924

---

- Instead of the old check `cstart < comment.finish`, I made a new one `finish >= comment.finish`.

The old one implied that the `comment.text` end could be more than the `comment.finish`, but was written with error. I don't think that happens. I found a more common case where `result` can be a multiline annotation or an alias, while `doc` is the first line, so `result.finish >>> comment.finish` and we can't parse comment. In any case `string.find` prevented the error, but I wanted to make it an explicit check.

- Fixed crash with a `/*` has no `.mark` field, it was type **comment.long** instead of **comment.clong**.
- Fixed typo `docGenric` -> `docGeneric`.
- Fix incorrect `rest.range` and `finish` on potentially multiline types.

- I added `%s*` to the **comment.long** pattern `^%s*@()` to support long string with leading spaces. But the string does not contain them due to a bug #3034.
```lua
--[===[    @param a string?ComMMS1]===]
local function new(a) end
```
This pattern is optional `*` and does not affect the current behavior.

tomlau10 commented on script/parser/luadoc.lua:
```diff
@@ -1725,42 +1724,47 @@ local function buildLuaDoc(comment)
+    local startOffset = comment.start + headPos
     if comment.type == 'comment.long' then
-        startOffset = startOffset + #comment.mark - 2
+        startOffset = comment.start + headPos + #comment.mark - 2
```
Seems that this line can be kept unchanged?
As the original line can be interpreted as `startOffset += #comment.mark - 2`

TIMONz1535 replied:
Yes, but **short** startOffset just simplified, actual is  `comment.start + headPos + 2 - 2` where the last +2 has nothing to do with the **long** one. So by logic you can't reuse **short** for **long**, but you can by math.

It seemed to me more transparent there than when we reuse the **short** for the **long**, as if it makes a special sense (eh... like short is a part of long?). And Lua's micro-optimizations don't excite me.

tomlau10 replied:
I'm not making this comment for micro optimization 🤔
It's that the current code is a bit confusing:
- There is already a `startOffset = comment.start + headPos`
- why the code below write the same part **again** but reuse the value?

---
> eh... like short is a part of long?

I see your point now 😮 though the maths can be simplified to have the same common part `comment.start + headPos`, their actual meaning is totally different.
If that's the case, I suggest using explicit `if / else` 🤔 
```lua
local startOffset
if comment.type == 'comment.short' then
    startOffset = comment.start + headPos
else
    startOffset = comment.start + headPos + #comment.mark - 2
end
```
=> this seemed to emphasis more that **different comment type** have **different formula** ?

tomlau10 commented on script/core/semantic-tokens.lua:
```diff
@@ -902,23 +902,23 @@ return function (uri, start, finish)
+                -- absolute position of `@` symbol
+                local startOffset = comm.start + headPos
+                if comm.type == 'comment.long' then
+                    startOffset = comm.start + headPos + #comm.mark - 2
```
this can also be written as `startOffset = startOffset + #comm.mark - 2`
=> to express `startOffset += #comm.mark - 2`

tomlau10 reviewed (commented):
Seems that your latest changes can solve #3034 as well?
Then you can **link** it by adding this to the first line of original your **PR description**
```md
fixes #3034 
```
https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

This can let github close the linked issues on this PR merge 👀

TIMONz1535 commented:
> your latest changes can solve #3034 as well?

No, this is a potential support if the bug is resolved in the future. But if is not, it does not affect anything.

Actually the pattern for **short** already contained optional spaces, but not for **long** ones. I did for the **long** ones as well.

TIMONz1535 commented:
@tomlau10 can I just add `mark = token` field to **comment.short, comment.cshort, comment.clong** and don't think about the type of comment anymore?

This will eliminate the ambiguity, and we will no longer keep the `--` prefix for **comment.short** in our head.

https://github.com/LuaLS/lua-language-server/blob/9ba27a6edd0ec0c73c3d59b3b1b028a1a46138c6/script/parser/compile.lua#L578-L597

tomlau10 commented:
> can I just add `mark = token` field to **comment.short, comment.cshort, comment.clong** and don't think about the type of comment anymore?

I don't think so 😕 
By looking at the code, the concept of **mark** is the **tokens** that are **quoting** the content/body.
For example there is also a **mark** concept for **string**, which is single quote or double quote `'` / `"`:
https://github.com/LuaLS/lua-language-server/blob/cb964c600570e6258d3c0a3f3f424a35a3a4ef64/script/parser/luadoc.lua#L697-L719

Back to comment, for short comment there is **no** mark concept, as nothing is used to quote the comment body.
v.s. long comment it is the `[[` / `[====[` that quoting the comment body.

---

Maybe you can just keep this PR as is.
Our latest discussion about the formula is not too meaning 🙈 
By [Law of triviality](https://en.wikipedia.org/wiki/Law_of_triviality) we should not focus too much on things that are too trivial.
